### DISCUSSION TITLE
Can I enable debugging mid-simulation?

### DISCUSSION THREAD

**Discussion by joshualmashburn** [gem5/gem5#1598] [Q&A] [answered]
Hello! I am working with Simpoint checkpoints I created. I am still new to the standard library, so I am basing my approach off of https://github.com/gem5/gem5/blob/stable/configs/example/gem5_library/checkpoints/simpoints-se-restore.py

I have added a custom debug flag, as I am trying to grab instruction operands and opcodes as they are sent to a certain execution unit (RISCV ISA). I have all that working easily, however, I do not want the warmup instructions included in the trace. The way I envision this is to enable debugging when the script schedules the actual Simpoint interval's instructions, such as on line 144 of simpoints-se-restore.py. 

I tried `m5.debug.flags['CustomFlag'] = True`, but I get the error `TypeError: 'AllFlags' object does not support item assignment`. Is there a canonical way to do this? I know the `--debug-time` argument exists, but the checkpoint generated is named after the tick at the start of the warmup, not the start of the Simpoint interval, and I have many of these to do, so I'd like to avoid manual intervention. Thanks!

**Comment by joshualmashburn** [comment, marked as answer]
I dug around in the m5 source code and found the code where the debug flags are parsed from the command line.
I tried `m5.debug.flags['CustomFlag'].enable()` and it works! I'm unsure if that's the proper way to do it, but if there's a better way, I'm curious to know.

**Reply by powerjg** [reply]
This is the right way!

**Comment by polpetras** [comment]
Hi,

I found this discussion after implementing exactly this feature for our own
simulation workflows. I have a working implementation and would like to
propose it as an upstream contribution.

**Design overview:**

- gem5 installs a handler on `SIGRTMIN+1` at startup (`SIGRTMIN` is reserved
  for KVM's `KVM_KICK_SIGNAL`, so +1 avoids the conflict).
- The signal handler follows the existing gem5 async pattern: it sets
  `async_event` + `async_debug_cmd` and wakes up the primary event queue. The actual
  command processing runs on the main simulation thread at the next event
  loop iteration — no mid-instruction races.
- Commands are written to `/tmp/gem5_debug_<pid>.cmd` before sending the
  signal. Supported commands: `+FLAG` (enable), `-FLAG` (disable),
  `+FLAG:Nt` (enable for N ticks then auto-disable), `trace:on`,
  `trace:off`.
- A companion shell script `util/gem5_debug_ctl.sh` handles writing the
  command file and sending the signal, so no C or Python tooling is needed
  on the user side.
- As of now, the feature is Linux-only — real-time signals are not available
  on macOS. The code is guarded with `#ifdef SIGRTMIN` and compiles cleanly
  on all platforms.
- `Flag::_tracing` is made `std::atomic<bool>` (relaxed ordering) to
  eliminate the data race that would otherwise exist when `changeFlag()` is
  called while other host threads are executing `DPRINTF`.

I will open a PR shortly. Happy to hear feedback on the design before or
during review.
